### DISCUSSION TITLE
How do I know if the token is expired or other errors?

### DISCUSSION THREAD

**Discussion by push-and-pop** [golang-jwt/jwt#133] [Q&A] [answered]
```
func ParseToken(token string) (*TokenClaim, error) {
	jwtToken, err := jwt.ParseWithClaims(token, &TokenClaim{}, func(t *jwt.Token) (interface{}, error) {

		return secret, nil
	})
	fmt.Println(err.Error())
	if err != nil {
		return nil, err
	}

	if claims, ok := jwtToken.Claims.(*TokenClaim); ok && jwtToken.Valid {
		return claims, nil
	} else {
		return nil, err
	}

}
```
I have a requirement, if it is expired, I want to return a new token, other errors will continue to make mistakes

**Comment by push-and-pop** [comment]
please A help

**Comment by 2yuri** [comment]
jwtToken.Valid return if token is valid or not

**Comment by oxisto** [comment, marked as answer]
You need to access our specific error type `ValidationError` and then can access the bit mask to get the type of error. We are working to improve this. The first step is to support error unwrapping with #125 but this does not work consistently for all claims yet. So you need to do the following, if `err` is the error you get from `Parse` or `ParseWithClaims`.

```golang
var ve *jwt.ValidationError
if errors.As(err, &ve) {
	if ve.Errors&jwt.ValidationErrorExpired != 0 {
		// Token is either expired or not active yet
		fmt.Println("Timing is everything")
	}
}
```

Similar code can be found in the error checking example https://github.com/golang-jwt/jwt/blob/c435f38291bfed5322cf0f4ed12d1f8668ceaeb3/example_test.go#L96-L119 (which we should update to support Go 1.13 error styles)
